### ISSUE TITLE
[Question] Is the reconfiguration of applications supported?

### ISSUE THREAD

**Issue by Mantany** [kube-hetzner/terraform-hcloud-kube-hetzner#363]
First of all: thank you for this great project!

I do however have a question: Is it possible to reconfigure the applications on a running cluster?
I build a working k3s-cluster with it, enabled cert-manager, but it seems when I try to disable it again, the changes do not get picked up. -> "No changes. Your infrastructure matches the configuration.". Also the same, when I try to add deployments like longhorn or want to switch from traefik to nginx.

**Comment by mysticaltech** [comment]
@Mantany Yes, it's one way for these applications via terraform. But those applications are configured via HelmChart definitions. You can use `kubectl get helmchart -n kube-system`, then you can do grab them with `kubectl get helmchart <name> -o yaml` same it to a file, and kubectl delete.

If you want a GUI for all of this, most probably Rancher is your best bet. I personally deploy Rancher as a single node (cx21) cluster with rancher enabled for this kind of things.

In the Readme, there's a lot of explanations on how HelmChart / HelmChart config work, please have a look.

**Comment by WolfspiritM** [comment]
Another way to modify and add stuff would be to use the created kustomization files and modify them.
There is a kustomize backup saved but it doesn't contain everything required.
However on the first control plane server you'll find a folder "/var/post_install/".
In there are all the files applied on the first install.

You can either change stuff in their and apply it via kustomize or copy it to your local machine and apply it from there.
What I did is moving this files to gitops (argocd) so they can be changed however needed.

Here is how it is applied on the first install so you can do the same:
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/5becc6f50fd264e46915652a6a64a6ea029278b7/init.tf#L265

If you want to add e.g. longhorn you can get the "longhorn.yaml" you need from the template here:
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/tree/master/templates

In there you can find a "${values}" placeholder. This placeholder is filled by either a supplied yaml file on the first setup or by some default values. The default values for that can be found here:
https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/blob/e466f0357d9505436594f816c5bfcd1d4b3aae79/locals.tf#L303

So for longhorn you would replace the "${values}" placeholder with the "default_longhorn_values" and use ext4 as fstype and 3 as replica count (or change it as you want).

Then you can either add that longhorn.yaml to the kustomization yaml and apply the kustomization or apply that file manually. That way you'd have longhorn deployed as it would have been on the first apply.

These "HelmChart" resources are just kubernetes resources that can be removed, modified and added while the k3s operator for helm charts makes sure they get installed as needed.

**Comment by Mantany** [comment]
Thank you for the great suggestions. 
In the past, I used a git repo and fluxCD for automatic deployment. That way I have the complete state of the cluster in one repo. I guess, for now, I will copy the kustomization files manually over. [Monorepo approach](https://fluxcd.io/flux/guides/repository-structure/)

Do you think think that it is useful to manage everything (longhorn, hetzner cloud manger, hetzner csi) via fluxCD? Do you have experience regarding bigger production clusters? 

That way I would only use terraform for the infrastructure and fluxCD deployment and fluxCD to manage the complete state of the cluster.

**Comment by mysticaltech** [comment]
@WolfspiritM 100% I had forgotten to mention the `/var/post_install` folder with all the kustomization on the first control plane.

@Mantany Definitely, I think you could turn off every extra and start fresh from FluxCD.
